### PR TITLE
[API][CatalogPromotion] refactor Variants Provider to strategy Pattern

### PR DESCRIPTION
**Pull request by AdamKasp**
| Q               | A
| --------------- | -----
| Branch?         | master
| Bug fix?        | no
| New feature?    | yes
| BC breaks?      | no
| License         | MIT

### CODE REVIEW

File: src/Sylius/Bundle/CoreBundle/Provider/Processable.php [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace Sylius\Bundle\CoreBundle\Provider;
+
+interface Processable
```
Comment (GSadee) [inline src/Sylius/Bundle/CoreBundle/Provider/Processable.php:16]:
```suggestion
interface VariantsProviderInterface
```
and there is missing `provideEligibleVariants` method

File: src/Sylius/Bundle/CoreBundle/Provider/VariantsBasedOnCatalogPromotionVariantsRuleProvider.php [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+namespace Sylius\Bundle\CoreBundle\Provider;
+
+use Sylius\Component\Core\Model\CatalogPromotionRuleInterface as CoreCatalogPromotionRuleInterface;
+use Sylius\Component\Core\Model\ProductVariantInterface;
+use Sylius\Component\Core\Repository\ProductVariantRepositoryInterface;
+use Sylius\Component\Promotion\Model\CatalogPromotionRuleInterface;
+use Webmozart\Assert\Assert;
+
+final class VariantsBasedOnCatalogPromotionVariantsRuleProvider implements Processable
```
Comment (GSadee) [inline src/Sylius/Bundle/CoreBundle/Provider/VariantsBasedOnCatalogPromotionVariantsRuleProvider.php:22]:
```suggestion
final class ForVariantsRuleVariantsProvider implements VariantsProviderInterface
```

File: src/Sylius/Bundle/CoreBundle/Provider/ForVariantsRuleVariantsProvider.php [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    public function provideEligibleVariants(CatalogPromotionRuleInterface $rule): array
+    {
+        $configuration = $rule->getConfiguration();
+
+        Assert::keyExists($configuration, 'variants', 'This rule should have configured variants');
+
+        return $this->getVariants($configuration['variants']);
+    }
+
+    private function getVariants(array $configuration): array
```
Comment (Zales0123) [inline src/Sylius/Bundle/CoreBundle/Provider/ForVariantsRuleVariantsProvider.php:44]: What about using a little bit of magic? 🧙 

```php
private function getVariants(array $configuration): array
{
    $variants = array_map(function(string $variantCode): ?ProductVariantInterface {
        return $this->productVariantRepository->findOneBy(['code' => $variantCode]);
    }, $configuration);

    return array_filter($variants, fn(?ProductVariantInterface $value) => $value !== null);
}
```

File: src/Sylius/Bundle/CoreBundle/spec/Provider/CatalogPromotionVariantsProviderSpec.php [hunk 4]
```diff
@@ -15,45 +15,37 @@
...
 
 final class CatalogPromotionVariantsProviderSpec extends ObjectBehavior
 {
-    function let(
-        ProductVariantRepositoryInterface $productVariantRepository
-    ): void {
-        $this->beConstructedWith($productVariantRepository);
+    function let(VariantsProviderInterface $provider): void
+    {
+        $this->beConstructedWith([$provider]);
```
Comment (Zales0123) [inline src/Sylius/Bundle/CoreBundle/spec/Provider/CatalogPromotionVariantsProviderSpec.php:28]: Let's use more than one provider here

File: src/Sylius/Bundle/CoreBundle/spec/Provider/ForVariantsRuleVariantsProviderSpec.php [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+        ProductVariantInterface $secondVariant
+    ): void {
+        $rule->getConfiguration()->willReturn(['variants' => ['PHP_T_SHIRT_XS_WHITE', 'PHP_T_SHIRT_XS_BLACK', 'PHP_MUG']]);
+
+        $productVariantRepository->findOneBy(['code' => 'PHP_T_SHIRT_XS_WHITE'])->willReturn($firstVariant);
+        $productVariantRepository->findOneBy(['code' => 'PHP_T_SHIRT_XS_BLACK'])->willReturn($secondVariant);
+        $productVariantRepository->findOneBy(['code' => 'PHP_MUG'])->willReturn(null);
+
+        $firstVariant->getCode()->willReturn('PHP_T_SHIRT_XS_WHITE');
+        $secondVariant->getCode()->willReturn('PHP_T_SHIRT_XS_BLACK');
```
Comment (Zales0123) [inline src/Sylius/Bundle/CoreBundle/spec/Provider/ForVariantsRuleVariantsProviderSpec.php:47]: I believe these two lines are not needed

File: src/Sylius/Bundle/CoreBundle/Provider/VariantsProviderInterface.php [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+
+declare(strict_types=1);
+
+namespace Sylius\Bundle\CoreBundle\Provider;
+
+use Sylius\Component\Core\Model\CatalogPromotionRuleInterface;
+
+interface VariantsProviderInterface
+{
+    public function supports(string $catalogPromotionRuleType): bool;
```
[the remark this answers is not in the document]
Comment (AdamKasp) [reply src/Sylius/Bundle/CoreBundle/Provider/VariantsProviderInterface.php:20]: Do you have some case when the entire CatalogPromotion can be necessary here? We want to iterate over all services which will implement this interface. and provide variant rule by rule, if someone wants to make some big customization then should customize `src/Sylius/Bundle/CoreBundle/Provider/CatalogPromotionVariantsProvider.php`

File: src/Sylius/Bundle/CoreBundle/Resources/config/services.xml [hunk 7]
```diff
@@ -265,7 +265,12 @@
             id="Sylius\Component\Core\Provider\CatalogPromotionVariantsProviderInterface"
             class="Sylius\Bundle\CoreBundle\Provider\CatalogPromotionVariantsProvider"
         >
+            <argument type="tagged_iterator" tag="sylius.catalog_promotion.variants_provider" />
+        </service>
+
+        <service id="Sylius\Bundle\CoreBundle\Provider\ForVariantsRuleVariantsProvider">
             <argument type="service" id="sylius.repository.product_variant" />
+            <tag name="sylius.catalog_promotion.variants_provider" />
```
Comment (Zales0123) [inline src/Sylius/Bundle/CoreBundle/Resources/config/services.xml:273]: Btw, we should probably mention this tag somewhere in the documentation 🖖

File: src/Sylius/Bundle/CoreBundle/spec/Provider/CatalogPromotionVariantsProviderSpec.php [hunk 8]
```diff
@@ -15,45 +15,37 @@
...
         $rule->getConfiguration()->willReturn(['variants' => ['PHP_T_SHIRT_XS_WHITE', 'PHP_T_SHIRT_XS_BLACK', 'PHP_MUG']]);
+        $rule->getType()->willReturn(CatalogPromotionRuleInterface::TYPE_FOR_VARIANTS);
 
-        $productVariantRepository->findOneBy(['code' => 'PHP_T_SHIRT_XS_WHITE'])->willReturn($firstVariant);
-        $productVariantRepository->findOneBy(['code' => 'PHP_T_SHIRT_XS_BLACK'])->willReturn($secondVariant);
-        $productVariantRepository->findOneBy(['code' => 'PHP_MUG'])->willReturn(null);
-
-        $firstVariant->getCode()->willReturn('PHP_T_SHIRT_XS_WHITE');
-        $secondVariant->getCode()->willReturn('PHP_T_SHIRT_XS_BLACK');
+        $firstProvider->supports($rule)->willReturn(true);
```
Comment (Zales0123) [inline src/Sylius/Bundle/CoreBundle/spec/Provider/CatalogPromotionVariantsProviderSpec.php:47]: If we return `false` here and `true` in the second one, we would also check if the if statement is done properly 💃

**Review by Zales0123 — APPROVED** [review]

File: src/Sylius/Bundle/CoreBundle/Resources/config/services.xml (hunk 7 above)
Comment (AdamKasp) [reply src/Sylius/Bundle/CoreBundle/Resources/config/services.xml:273]: absolutely we should, but IMHO it should be in separate PR (with an explanation of how it works with more rules), I will add subtask to our board :)

**Comment by Zales0123** [comment]
Thanks, Adam! :1st_place_medal:

**Pull request merged**